database/bfgd/database_ext_test.go

```diff
@@ -1653,6 +1653,168 @@ func TestL2BtcFinalitiesByL2KeystoneNotPublishedHeight(t *testing.T) {
 	}
 }
 
+func TestBtcHeightsNoChildren(t *testing.T) {
+	type testTableItem struct {
+		name                         string
+		numberToCreateWithChildren   int
+		numberToCreateWithNoChildren int
+		overlapCount                 int
+	}
+
+	testTable := []testTableItem{
+		{
+			name:                         "0",
+			numberToCreateWithNoChildren: 0,
+			numberToCreateWithChildren:   43,
+		},
+		{
+			name:                         "less than 100",
+			numberToCreateWithNoChildren: 76,
+			numberToCreateWithChildren:   4,
+		},
+		{
+			name:                         "more than 100",
+			numberToCreateWithNoChildren: 126,
+			numberToCreateWithChildren:   333,
+		},
+		{
+			name:                         "more than 100 and overlap",
+			numberToCreateWithNoChildren: 126,
+			numberToCreateWithChildren:   333,
+			overlapCount:                 98,
+		},
+	}
+
+	createBlocksWithNoChildren := func(ctx context.Context, count int, db bfgd.Database) []int64 {
+		heights := make([]int64, count)
+		for i := range count {
+			height := mathrand.Int64()
+			hash := make([]byte, 32)
+			if _, err := rand.Read(hash); err != nil {
+				t.Fatal(err)
+			}
+			header := make([]byte, 80)
+			if _, err := rand.Read(header); err != nil {
+				t.Fatal(err)
+			}
```

Comment:
As more of a note for later, as these are completely random there is no guarantee that the header won't have a previous hash that matches an existing hash, which could cause unexpected behaviour.

This would be extremely rare, but could cause test failures so I think it's worth noting.

Reply:
that's true, let me see if I can handle 🤔

Reply:
I think the fix you added checks whether a block's header contains its own "hash" as the previous hash, correct me if I am wrong.

I don't think this will prevent the issue I mentioned above, where another block's header will have the "previous hash" of another generated block "hash", which would make the block appear as so it is a child of another generated block.

Reply:
🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦 🤦